src/main/java/com/github/firmwehr/gentle/firm/optimization/ArithmeticOptimization.java

```diff
@@ -63,125 +69,160 @@ private void applyArithmeticOptimization() {
 	}
 
 	@Override
-	public void visit(Add node) {
-		if (tarValOf(node.getRight()).isNull()) {
-			exchange(node, node.getLeft());
-			return;
-		}
-		if (tarValOf(node.getLeft()).isNull()) {
-			exchange(node, node.getRight());
-		}
-	}
+	public void defaultVisit(Node node) {
+		// TODO (maybe): 2 * a => a << 2
+		timesOne(node).ifPresent(match -> exchange(node, match.other()));
+		timesZero(node).ifPresent(match -> exchange(node, node.getGraph().newConst(0, node.getMode())));
```

Comment:
You're calling `exchange(node, ...)` here but e.g. `exchange(match.add(), ...)` below, I think that should be more consistent as long as there isn't a reason I'm missing.

Reply:
`match.xxx` is well typed and `node` has fewer characters :P I can migrate all of them to `match.x`